File: common/sbis-common/src/main/java/ru/tensor/sbis/common/util/BindingAdapters.java
package ru.tensor.sbis.common.util;

import android.util.TypedValue;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.mikepenz.iconics.typeface.IIcon;
import com.mikepenz.iconics.view.IconicsImageView;

import androidx.annotation.ColorInt;
import androidx.annotation.ColorRes;
import androidx.annotation.DimenRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;
import androidx.databinding.BindingAdapter;
import kotlin.jvm.functions.Function1;
import ru.tensor.sbis.design.SbisMobileIcon;
import ru.tensor.sbis.design.utils.FormatUtils;
import ru.tensor.sbis.design.utils.iconics.IconicUtils;

public class BindingAdapters {

    @BindingAdapter("onClick")
    public static void onClick(@NonNull View view, @Nullable Runnable runnable) {
        view.setOnClickListener(v -> {
            if (runnable != null) {
                runnable.run();
            }
        });
    }

    @BindingAdapter("onLongClick")
    public static void onLongClick(@NonNull View view, @NonNull Runnable runnable) {
        view.setOnLongClickListener(v -> {
            runnable.run();
            return true;
        });
    }

    @BindingAdapter("textColorRes")
    public static void setTextColorRes(@NonNull TextView view,
                                       @ColorRes int colorResId) {
        if (colorResId != 0) {
            view.setTextColor(ContextCompat.getColor(view.getContext(), colorResId));
        }
    }

    @BindingAdapter("textSizeRes")
    public static void setTextSizeRes(@NonNull TextView view,
                                      @DimenRes int sizeResId) {
        if (sizeResId != 0) {
            view.setTextSize(TypedValue.COMPLEX_UNIT_PX, view.getContext().getResources().getDimension(sizeResId));
        }
    }

    @BindingAdapter({"highlightedText", "linkColor"})
    public static void setHighlightedText(@NonNull TextView view, @NonNull String text, @ColorInt int linkColor) {
        view.setText(TextFormatUtils.highlightLinks(text, linkColor));
    }

    @BindingAdapter({"highlightedText", "linkColor", "onUrlClick"})
    public static void setHighlightedText(@NonNull TextView view, @NonNull String text, @ColorInt int linkColor, Function1<String, Boolean> urlClickHandler) {
        view.setText(TextFormatUtils.highlightLinks(text, linkColor, urlClickHandler));
    }

    @BindingAdapter("layout_marginLeft")
    public static void setMarginLeft(@NonNull View view,
                                     @DimenRes int marginLeftResId) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        lp.leftMargin = (int)view.getContext().getResources().getDimension(marginLeftResId);
        view.setLayoutParams(lp);
    }

    @BindingAdapter("layout_marginRight")
    public static void setMarginRight(@NonNull View view,
                                      @DimenRes int marginRightResId) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        lp.rightMargin = (int)view.getContext().getResources().getDimension(marginRightResId);
        view.setLayoutParams(lp);
    }

    @BindingAdapter("layout_marginTop")
    public static void setMarginTop(@NonNull View view,
                                    @DimenRes int marginTopResId) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        lp.topMargin = (int) view.getContext().getResources().getDimension(marginTopResId);
        view.setLayoutParams(lp);
    }

    @BindingAdapter("layout_marginBottom")
    public static void setMarginBottom(@NonNull View view,
                                       @DimenRes int marginBottomResId) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        lp.bottomMargin = (int) view.getContext().getResources().getDimension(marginBottomResId);
        view.setLayoutParams(lp);
    }

    @BindingAdapter("layout_marginHorizontal")
    public static void setMarginHorizontal(@NonNull View view,
                                           @DimenRes int marginHorizontalResId) {
        ViewGroup.MarginLayoutParams lp = (ViewGroup.MarginLayoutParams) view.getLayoutParams();
        int horizontalMargin = (int) view.getContext().getResources().getDimension(marginHorizontalResId);
        lp.leftMargin = horizontalMargin;
        lp.rightMargin = horizontalMargin;
        view.setLayoutParams(lp);
    }

    @BindingAdapter("iconicsImageViewColor")
    public static void photoCollectionData(@NonNull IconicsImageView iconicsImageView, @ColorRes int color) {
        IconicUtils.setColorRes(iconicsImageView, color);
    }

    @BindingAdapter("iconicsImageViewIcon")
    public static void iconicsImageViewIconString(@NonNull IconicsImageView iconicsImageView, @NonNull String icon) {
        IconicUtils.setIcon(iconicsImageView, icon);
    }

    @BindingAdapter("iconicsImageViewIcon")
    public static void iconicsImageViewIIcon(@NonNull IconicsImageView iconicsImageView, @NonNull IIcon icon) {
        IconicUtils.setIcon(iconicsImageView, icon);
    }

    @BindingAdapter("visibilityEqualsGone")
    public static void isVisibilityOrGone(View view, boolean visible) {
        view.setVisibility(visible ? View.VISIBLE : View.GONE);
    }

    @BindingAdapter({"isVisible", "delay"})
    public static void isVisibleDelayed(View view, boolean visible, int delay) {
        final Runnable hideCallback = () -> view.setVisibility(View.INVISIBLE);
        view.removeCallbacks(hideCallback);
        if (visible) {
            view.setVisibility(View.VISIBLE);
        } else {
            view.postDelayed(hideCallback, delay);
        }
    }

    @BindingAdapter("invisibleOrGone")
    public static void isInvisibleOrGone(View view, boolean invisible) {
        view.setVisibility(invisible ? View.INVISIBLE : View.GONE);
    }

    @BindingAdapter("mobileIcon")
    public static void setMobileIcon(@NonNull TextView textView, @NonNull SbisMobileIcon.Icon icon) {
        textView.setText(String.valueOf(icon.getCharacter()));
    }

    @BindingAdapter("resId")
    public static void setTextRes(@NonNull TextView view,
                                  @StringRes int text) {
        if (text != 0) {
            view.setText(text);
        } else {
            view.setText("");
        }
    }

    @BindingAdapter("fontStyle")
    public static void setFontStyle(@NonNull TextView textView, int fontStyle) {
        textView.setTypeface(null, fontStyle);
    }

    @BindingAdapter("textVisibility")
    public static void setTextVisibility(@NonNull TextView view, String text) {
        view.setVisibility((text == null || text.isEmpty()) ? View.GONE : View.VISIBLE);
    }

    @BindingAdapter("formattedCount")
    public static void setFormattedCount(@NonNull TextView view, int count) {
        view.setText(FormatUtils.formatCount(count));
    }

    @BindingAdapter("selected")
    public static void setSelected(@NonNull View view,
                                   boolean selected) {
        view.setSelected(selected);
    }

    @BindingAdapter({"folderLevel", "paddingForLevel"})
    public static void paddingForLevel(@NonNull View view,
                                       int level,
                                       float paddingForLevel) {
        view.setPadding((int) (level * paddingForLevel),
                view.getPaddingTop(),
                view.getPaddingRight(),
                view.getPaddingBottom());
    }
}

File: design/design_sbis_text_view/src/main/java/ru/tensor/sbis/design/sbis_text_view/BindingAdapters.kt
package ru.tensor.sbis.design.sbis_text_view

import android.util.TypedValue
import android.view.View
import androidx.annotation.AttrRes
import androidx.annotation.ColorRes
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.core.view.isVisible
import androidx.databinding.BindingAdapter
import ru.tensor.sbis.design.SbisMobileIcon
import ru.tensor.sbis.design.utils.getDimen

/**
 * Биндинг-адаптеры для компонента [SbisTextView].
 *
 * @author vv.chekurda
 */
object BindingAdapters {

    /**
     * Установить цвет текста для [view] - SbisTextView при помощи ссылки на атрибут [attrResId].
     */
    @JvmStatic
    @BindingAdapter("textColorAttr")
    fun setTextColorAttr(view: SbisTextView, @AttrRes attrResId: Int) {
        if (attrResId != 0) {
            val typedValue = TypedValue()
            val theme = view.context.theme
            theme.resolveAttribute(attrResId, typedValue, true)
            view.setTextColor(typedValue.data)
        }
    }

    /**
     * Установить размер текста для [SbisTextView] при помощи ссылки на атрибут [attrResId].
     */
    @JvmStatic
    @BindingAdapter("textSizeAttr")
    fun SbisTextView.setTextSizeAttr(@AttrRes attrResId: Int) {
        if (attrResId != 0) {
            setTextSize(TypedValue.COMPLEX_UNIT_PX, context.getDimen(attrResId))
        }
    }

    /**
     * Установить стиль шрифта для [textView] - SbisTextView при помощи ресурса стиля [fontStyle].
     */
    @JvmStatic
    @BindingAdapter("fontStyle")
    fun setFontStyle(textView: SbisTextView, fontStyle: Int) {
        textView.setTypeface(null, fontStyle)
    }

    /**
     * Установить цвет текста для [view] - SbisTextView при помощи ресурса цвета [colorResId].
     */
    @JvmStatic
    @BindingAdapter("textColorRes")
    fun setTextColorRes(
        view: SbisTextView,
        @ColorRes colorResId: Int
    ) {
        if (colorResId != 0) {
            view.setTextColor(ContextCompat.getColor(view.context, colorResId))
        }
    }

    /**
     * Устанавливает видимость SbisTextView в зависимости от текста.
     * Если text=null, то видимость будет установлена в View.GONE
     */
    @JvmStatic
    @BindingAdapter("visibleOrGone")
    fun SbisTextView.visibleOrGone(text: String?) {
        visibility = if (text.isNullOrEmpty()) View.GONE else View.VISIBLE
    }

    /**@SelfDocumented */
    @JvmStatic
    @BindingAdapter("textSizeRes")
    fun SbisTextView.textSizeRes(resId: Int) = setTextSize(
        TypedValue.COMPLEX_UNIT_PX,
        resources.getDimension(resId)
    )

    /**@SelfDocumented */
    @JvmStatic
    @BindingAdapter("textRes")
    fun SbisTextView.setTextRes(@StringRes text: Int) {
        setText(text)
    }

    /**
     * Установить текст в SbisTextView при помощи строкового ресусра [text], если text=0 спрятать.
     */
    @JvmStatic
    @BindingAdapter("textResOrGone")
    fun SbisTextView.setTextResOrGone(@StringRes text: Int) {
        if (text != 0) {
            setText(text)
        }
        isVisible = text != 0
    }

    /**
     * Установить текст в SbisTextView при помощи последовательности символов [text], если text=null спрятать.
     */
    @JvmStatic
    @BindingAdapter("textAndVisibility")
    fun SbisTextView.setTextAndVisibility(text: CharSequence?) {
        this.text = text
        val visible = text != null && text.isNotEmpty()
        visibility = if (visible) View.VISIBLE else View.GONE
    }

    /**
     * Установить подсветку текста [text] исходя из [highlightedRanges].
     * Адаптер создан для поддержки внедрения SbisTextView.
     * Оригинальный адаптер по пути [common/common-business/src/main/java/ru/tensor/sbis/business/common/ui/bind_adapter/BindingAdapters.kt]
     */
    @JvmStatic
    @BindingAdapter("textWithHighlightedRanges", "highlightedRanges")
    fun SbisTextView.setTextWithHighlightedRanges(
        text: String,
        highlightedRanges: List<IntRange>
    ) {
        setTextWithHighlightRanges(text, highlightedRanges)
    }

    /**
     * Установить иконку [icon] в SbisTextView.
     */
    @JvmStatic
    @BindingAdapter("mobileIcon")
    fun SbisTextView.setMobileIcon(icon: SbisMobileIcon.Icon) {
        text = icon.character.toString()
    }

}
